compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/writer/BIRInstructionWriter.java

```diff
@@ -0,0 +1,606 @@
+import org.wso2.ballerinalang.compiler.bir.writer.CPEntry.ByteCPEntry;
+import org.wso2.ballerinalang.compiler.bir.writer.CPEntry.FloatCPEntry;
+import org.wso2.ballerinalang.compiler.bir.writer.CPEntry.IntegerCPEntry;
+import org.wso2.ballerinalang.compiler.bir.writer.CPEntry.StringCPEntry;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BType;
+import org.wso2.ballerinalang.compiler.util.TypeTags;
+import org.wso2.ballerinalang.compiler.util.diagnotic.DiagnosticPos;
+
+import java.util.List;
+
+/**
+ * Responsible for serializing BIR instructions and operands.
+ *
+ * @since 0.980.0
+ */
+public class BIRInstructionWriter extends BIRVisitor {
+    private ByteBuf buf;
+    private BIRTypeWriter typeWriter;
+    private ConstantPool cp;
+    private BIRBinaryWriter binaryWriter;
+
+    public BIRInstructionWriter(ByteBuf buf, BIRTypeWriter typeWriter, ConstantPool cp,
+                                BIRBinaryWriter birBinaryWriter) {
+        this.buf = buf;
+        this.typeWriter = typeWriter;
+        this.binaryWriter = birBinaryWriter;
+        this.cp = cp;
+    }
+
+    public void writeBBs(List<BIRBasicBlock> bbList) {
+        buf.writeInt(bbList.size());
+        bbList.forEach(bb -> bb.accept(this));
+    }
+
+    public void visit(BIRBasicBlock birBasicBlock) {
+        //Name of the basic block
+        addCpAndWriteString(birBasicBlock.id.value);
+        // Number of instructions
+        // Adding the terminator instruction as well.
+        buf.writeInt(birBasicBlock.instructions.size() + 1);
+            birBasicBlock.instructions.forEach(instruction -> instruction.accept(this));
+        if (birBasicBlock.terminator == null) {
+            throw new BLangCompilerException("Basic block without a terminator : " + birBasicBlock.id);
+        }
+        birBasicBlock.terminator.accept(this);
+    }
+
+    public void writeErrorTable(List<BIRNode.BIRErrorEntry> errorEntries) {
+        buf.writeInt(errorEntries.size());
+        errorEntries.forEach(birErrorEntry -> birErrorEntry.accept(this));
+    }
+
+    public void visit(BIRNode.BIRErrorEntry errorEntry) {
+        addCpAndWriteString(errorEntry.trapBB.id.value);
+        addCpAndWriteString(errorEntry.endBB.id.value);
+        errorEntry.errorOp.accept(this);
+        addCpAndWriteString(errorEntry.targetBB.id.value);
+    }
+
+    // Terminating instructions
+
+    public void visit(BIRTerminator.GOTO birGoto) {
+        writePosition(birGoto.pos);
+        buf.writeByte(birGoto.kind.getValue());
+        addCpAndWriteString(birGoto.targetBB.id.value);
+    }
+
+    public void visit(BIRTerminator.Lock lock) {
+        writePosition(lock.pos);
+        buf.writeByte(lock.kind.getValue());
+
+        addCpAndWriteString(lock.lockedBB.id.value);
+    }
+
+    public void visit(BIRTerminator.FieldLock lock) {
+        writePosition(lock.pos);
+        buf.writeByte(lock.kind.getValue());
+        // TODO properly use operand instead of variablDcl.name here
```

Comment:
shall we create an issue to track the todo indicating how to properly use the operand here?